p2p/proto/state.proto

```diff
@@ -0,0 +1,39 @@
+syntax = "proto3";
+import "p2p/proto/common.proto";
+
+
+// optimized for flat storage, not through a trie (not sharing key prefixes)
+message ContractStoredValue {
+    Felt252 key = 1;
+    Felt252 value = 2;
+}
+
+message StateDiff
+{
+    // a bit more efficient than the state sync separation
+    message ContractDiff {
+        Address address = 1;
+        Felt252 nonce = 2;
+        Felt252 class_hash = 3; // can change for replace_class or new contract
+        repeated ContractStoredValue values = 4;
+    }
+
+    uint32 tree_id = 1;
+    repeated ContractDiff contract_diffs = 2;
+}
+
+// is it better to separate the definition from the hashes? (will need to repeate the hashes
+// for the definitions stream)
+// or, make the definitions optional? maybe it is enough to know only that a class exists, not its definition
+// which may be fetched lazily later.
+message Class {
```

Comment:
What is the intended way of fetching class definitions in P2P? 
I dont see this being used outside of `snapshot.proto`

Reply:
Sent as state diffs which are considered the block body.

Reply:
Can you point me to the definition of that, I cant seem to figure out which one exactly you are referring to.

Reply:
> // result is (BlockHeader, StateDiff*)*
> message GetBlocks {

If you have a suggestion on how to document results better, I'm eager to hear it.

Reply:
I'm afraid that due to size of the definitions it would be better to not include them directly in the StateDiff but rather stream them along the rest of the state diff:
```
// result is (BlockHeader, StateDiff*, Classes*, Class*)*
message GetBlocks {
```
or request them separately/lazily.

Reply:
I prefer having them as part of the response, not separate, unless you see a use case where a client just wants the state but doesn't care about classes.

Reply:
Classes were supposed to be part of the GetBlocks response, as noted in notes.md, but not copy-pasted to the comment above GetBlocks...

Reply:
Sorry, I forgot about `notes.md`.

So as per slack discussion:
>GetBlocksResponse is either BH or a pair of SD and Block ID, and the return value is GetBlocksResponse*

if GetBlocksResponse wrapper is added then the response to GetBlocks is GetBlocksResponse*, where the possible variants are:
GetBlocksResponse(BlockHeader)
GetBlocksResponse(StateDiff, BlockID)
GetBlocksResponse(Classes, BlockID)

What about signatures and block proof?

Reply:
Will be included as well.